crates/dbsp/src/circuit/circuit_builder.rs

```diff
@@ -564,327 +574,978 @@ impl Edge {
+    ///
+    /// Normally a strict operator writes a value computed based on inputs from
+    /// previous clock cycles to its output stream at the start of each new
+    /// clock cycle.  When the local clock epoch ends, the last value
+    /// computed by the operator (that would otherwise be dropped) is
+    /// written to the export stream instead.
+    ///
+    /// # Examples
+    ///
+    /// See example in the [`Self::iterate`] method.
+    fn add_feedback_with_export<I, O, Op>(
+        &self,
+        operator: Op,
+    ) -> (ExportStream<Self, O>, FeedbackConnector<Self, I, O, Op>)
+    where
+        I: Data,
+        O: Data,
+        Op: StrictUnaryOperator<I, O>;
+
+    fn connect_feedback_with_preference<I, O, Op>(
+        &self,
+        output_node_id: NodeId,
+        operator: Rc<UnsafeCell<Op>>,
+        input_stream: &Stream<Self, I>,
+        input_preference: OwnershipPreference,
+    ) where
+        I: Data,
+        O: Data,
+        Op: StrictUnaryOperator<I, O>;
+
+    /// Add a child circuit.
+    ///
+    /// Creates an empty circuit with `self` as parent and invokes
+    /// `child_constructor` to populate the circuit.  `child_constructor`
+    /// typically captures some of the streams in `self` and connects them
+    /// to source nodes of the child circuit.  It is also responsible for
+    /// attaching an executor to the child circuit.  The return type `T`
+    /// will typically contain output streams of the child.
+    ///
+    /// Most users should invoke higher-level APIs like [`Circuit::iterate`]
+    /// instead of using this method directly.
+    fn subcircuit<F, T, E>(
+        &self,
+        iterative: bool,
+        child_constructor: F,
+    ) -> Result<T, SchedulerError>
+    where
+        F: FnOnce(&mut ChildCircuit<Self>) -> Result<(T, E), SchedulerError>,
+        E: Executor<ChildCircuit<Self>>;
+
+    /// Add an iteratively scheduled child circuit.
+    ///
+    /// Add a child circuit with a nested clock.  The child will execute
+    /// multiple times for each parent timestamp, until its termination
+    /// condition is satisfied.  Every time the child circuit is activated
+    /// by the parent (once per parent timestamp), the executor calls
+    /// [`clock_start`](`super::operator_traits::Operator::clock_start`)
+    /// on each child operator.  It then calls `eval` on all
+    /// child operators in a causal order and checks if the termination
+    /// condition is satisfied.  If the condition is `false`, the
+    /// executor `eval`s all operators again.  Once the termination
+    /// condition is `true`, the executor calls `clock_end` on all child
+    /// operators and returns control back to the parent scheduler.
+    ///
+    /// The `constructor` closure populates the child circuit and returns a
+    /// closure that checks the termination condition and an arbitrary
+    /// user-defined return value that typically contains output streams
```

Comment:
Does this mean that it can check any of the streams in the circuit is it constructing? "return value" is confusing.

Reply:
In fact, the example below looks at more than just streams.

Reply:
Yes, this is a low-level function that can be used to build all kinds of weird termination conditions.  I don't expect people to use it directly.